File: 作业1-通用分布函数/curve_fit.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm, beta
from scipy.optimize import curve_fit

def fit_distribution(bin_num):
    csv_name = f'Pw_bin_csv/Pw_bin_{bin_num}.csv'
    cdf_name = f'cdf_fit_png/cdf_fit_{bin_num}.png'
    pdf_name = f'pdf_fit_png/pdf_fit_{bin_num}.png'
    txt_name = f'para_fit_txt/para_fit_{bin_num}.txt'

    # 加载数据
    data = np.array(pd.read_csv(csv_name)['value'].values)
    # 去掉为0的值（风电功率为0时影响bete分布拟合）
    # data = data[data > 0]
    data_sorted = np.sort(data)
    y_actual_cdf = np.linspace(0, 1, len(data))
    x_range = np.linspace(0, 1, 500)

    # ---------------------------------------------------------
    # 2. 定义分布模型
    # ---------------------------------------------------------

    # Versatile CDF 
    def safe_exp(z):
        return np.exp(np.clip(z, -700, 700))

    def versatile_cdf(x, a, b, c):
        exp_term = safe_exp(-a * (x - c))
        return (1 + exp_term)**(-b)

    # Versatile PDF: CDF的导数 
    def versatile_pdf(x, a, b, c):
        exp_term = safe_exp(-a * (x - c))
        return (a * b * exp_term) / ((1 + exp_term)**(b + 1))

    # ---------------------------------------------------------
    # 3. 拟合计算
    # ---------------------------------------------------------

    # 高斯分布拟合 (MLE)
    mu, std = norm.fit(data)

    # Beta 分布拟合 (MLE)
    a_beta, b_beta, loc_beta, scale_beta = beta.fit(data, floc=0, fscale=1)

    # Versatile 分布拟合 (基于 CDF 曲线拟合)
    # 初始值参考论文 Table I: a=15, b=1.2, c=0.4
    # 使用参数 bounds 防止搜索到不合理区域，并限制 exp 数值范围以避免 overflow
    p0 = [15, 1.2, 0.4]
    lower = [0.01, 0.01, 0.0]
    upper = [1000.0, 10.0, 1.0]
    try:
        popt, _ = curve_fit(versatile_cdf, data_sorted, y_actual_cdf, p0=p0, bounds=(lower, upper), maxfev=5000)
        a_v, b_v, c_v = popt
    except RuntimeError:
        # 如果首次拟合未收敛，重试：更小的初始 a、更多迭代次数
        try:
            popt, _ = curve_fit(versatile_cdf, data_sorted, y_actual_cdf, p0=[5.0, 1.0, 0.2], bounds=(lower, upper), maxfev=20000)
            a_v, b_v, c_v = popt
        except Exception as e:
            print(f"Versatile fit failed for bin {bin_num}: {e}")
            # 退回到论文参考值，避免中断后续绘图
            a_v, b_v, c_v = (15.0, 1.2, 0.4)

    # ---------------------------------------------------------
    # 4. 绘制 CDF 曲线 (类似论文 Fig. 3)
    # ---------------------------------------------------------
    plt.figure(figsize=(10, 8))
    # 绘制散点 (注意步长)
    plt.scatter(data_sorted[::20], y_actual_cdf[::20], label='Actual', color='black', marker='o', facecolors='none')
    plt.plot(x_range, norm.cdf(x_range, mu, std), label='Gaussian', linestyle='-')
    plt.plot(x_range, beta.cdf(x_range, a_beta, b_beta, loc_beta, scale_beta), label='Beta', linestyle='-')
    plt.plot(x_range, versatile_cdf(x_range, a_v, b_v, c_v), label='Versatile', linestyle='--', color='red', linewidth=2)
    plt.xlabel('Probabilistic wind power output')
    plt.ylabel('CDF')
    plt.legend()
    plt.title('Comparison of CDF Fitting')
    plt.grid(True, linestyle=':', alpha=0.6)
    plt.savefig(cdf_name, dpi=300)
    plt.close()

    # ---------------------------------------------------------
    # 5. 绘制 PDF 曲线 (类似论文 Fig. 1 & 4)
    # ---------------------------------------------------------
    plt.figure(figsize=(10, 8))

    # 计算原始数据概率密度散点 (40个点, 步长0.025) 
    counts, bin_edges = np.histogram(data, bins=100, range=(0, 1), density=True)
    bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2
    plt.scatter(bin_centers, counts, label='Actual (Binned)', color='black', marker='o', facecolors='none')

    # 计算各分布的 PDF 曲线
    pdf_gaussian = norm.pdf(x_range, mu, std)
    pdf_beta = beta.pdf(x_range, a_beta, b_beta, loc_beta, scale_beta)
    pdf_versatile = versatile_pdf(x_range, a_v, b_v, c_v)

    plt.plot(x_range, pdf_gaussian, label='Gaussian', linestyle='-')
    plt.plot(x_range, pdf_beta, label='Beta', linestyle='-')
    plt.plot(x_range, pdf_versatile, label='Versatile', linestyle='--', color='red', linewidth=2)

    plt.xlabel('Probabilistic wind power output')
    plt.ylabel('PDF')
    plt.legend()
    plt.title('Comparison of PDF Fitting')
    plt.grid(True, linestyle=':', alpha=0.6)
    plt.savefig(pdf_name, dpi=300)
    plt.close()

    # ---------------------------------------------------------
    # 6. 输出参数与 RMSE 评估 (公式 3)
    # ---------------------------------------------------------
    n_o = 100
    eval_points = np.linspace(1/n_o, 1, n_o) 
    actual_cdf_at_eval = np.interp(eval_points, data_sorted, y_actual_cdf)

    def get_rmse(sim_vals):
        return np.sqrt(np.mean((actual_cdf_at_eval - sim_vals)**2))

    rmse_g = get_rmse(norm.cdf(eval_points, mu, std))
    rmse_b = get_rmse(beta.cdf(eval_points, a_beta, b_beta, loc_beta, scale_beta))
    rmse_v = get_rmse(versatile_cdf(eval_points, a_v, b_v, c_v))

    with open(txt_name, 'w') as f:
        f.write('--- Fitting Parameters ---\n')
        f.write(f'Gaussian: mu={mu:.4f}, std={std:.4f}\n')
        f.write(f'Beta: a={a_beta:.4f}, b={b_beta:.4f}\n')
        f.write(f'Versatile: a={a_v:.4f}, b={b_v:.4f}, c={c_v:.4f}\n\n')
        f.write('--- RMSE Evaluation ---\n')
        f.write(f'Gaussian RMSE:  {rmse_g:.5f}\n')
        f.write(f'Beta RMSE:      {rmse_b:.5f}\n')
        f.write(f'Versatile RMSE: {rmse_v:.5f} (Optimal)\n')
    print(f'Bin {bin_num}: finished fitting.')

bin_nums = [f"{i:02d}" for i in range(1, 26)]
for bin_num in bin_nums:
    fit_distribution(bin_num)
print(f"任务完成！已生成：\n- CDF图: cdf_fit_png\n- PDF图: pdf_fit_png\n- 参数表: para_fit_txt")

File: 作业1-通用分布函数/curve_fit.m
%% 风电功率分布拟合 MATLAB 版
clear; clc; close all;

bin_num = "04"; % "12"
fit_single_bin(bin_num);

fprintf('任务完成！已生成：\n- CDF图: cdf_fit_png\n- PDF图: pdf_fit_png\n- 参数表: para_fit_txt\n');

%% 核心拟合函数
function fit_single_bin(bin_num)
    csv_name = fullfile(['Pw_bin_', bin_num, '.csv']);

    % --- 加载数据 ---
    opts = detectImportOptions(csv_name);
    T = readtable(csv_name, opts);
    data = T.value;
    data = data(~isnan(data)); % 去除NaN
    data_sorted = sort(data);
    
    n = length(data_sorted);
    y_actual_cdf = linspace(0, 1, n)';
    x_range = linspace(0, 1, 500)';

    % --- 定义分布模型 ---
    % Versatile CDF: (1 + exp(-a*(x-c)))^-b
    % p(1)=a, p(2)=b, p(3)=c
    versatile_cdf_func = @(p, x) (1 + exp(-min(max(-p(1)*(x-p(3)), -700), 700))).^(-p(2));
    
    % Versatile PDF
    versatile_pdf_func = @(p, x) (p(1)*p(2)*exp(-min(max(-p(1)*(x-p(3)), -700), 700))) ./ ...
                         ((1 + exp(-min(max(-p(1)*(x-p(3)), -700), 700))).^(p(2)+1));

    % --- 拟合计算 ---
    
    % 1. 高斯分布 (MLE)
    [mu, std_dev] = normfit(data);

    % 2. Beta 分布 (MLE)
    % 注意：MATLAB betafit 要求数据在 (0,1) 之间，进行极小值修正
    data_beta = data;
    data_beta(data_beta <= 0) = 0.0001;
    data_beta(data_beta >= 1) = 0.9999;
    beta_paras = betafit(data_beta);
    a_beta = beta_paras(1); b_beta = beta_paras(2);

    % 3. Versatile 分布拟合 (基于非线性最小二乘)
    p0 = [15, 1.2, 0.4];
    lb = [0.01, 0.01, 0.0];
    ub = [1000.0, 10.0, 1.0];
    
    options = optimset('Display','off', 'MaxIter', 5000, 'MaxFunEvals', 10000);
    try
        popt = lsqcurvefit(versatile_cdf_func, p0, data_sorted, y_actual_cdf, lb, ub, options);
    catch
        % 重试逻辑
        p0_retry = [5.0, 1.0, 0.2];
        popt = lsqcurvefit(versatile_cdf_func, p0_retry, data_sorted, y_actual_cdf, lb, ub, options);
    end
    a_v = popt(1); b_v = popt(2); c_v = popt(3);

    % --- 绘制 CDF 曲线 ---
    figure('Visible', 'off'); hold on;
    % 采样绘制散点以防过密
    idx = 1:floor(n/40):n;
    scatter(data_sorted(idx), y_actual_cdf(idx), 'k', 'LineWidth', 1);
    plot(x_range, normcdf(x_range, mu, std_dev), 'LineWidth', 1.5);
    plot(x_range, betacdf(x_range, a_beta, b_beta), 'LineWidth', 1.5);
    plot(x_range, versatile_cdf_func(popt, x_range), 'r--', 'LineWidth', 2);
    
    xlabel('Probabilistic wind power output'); ylabel('CDF');
    legend('Actual', 'Gaussian', 'Beta', 'Versatile', 'Location', 'Best');
    title(['Comparison of CDF Fitting - Bin ', bin_num]);
    grid on;
    saveas(gcf, cdf_name);
    close(gcf);

    % --- 绘制 PDF 曲线 ---
    figure('Visible', 'off'); hold on;
    [counts, edges] = histcounts(data, 100, 'BinLimits', [0, 1], 'Normalization', 'pdf');
    centers = (edges(1:end-1) + edges(2:end)) / 2;
    scatter(centers, counts, 'k', 'LineWidth', 1);
    
    plot(x_range, normpdf(x_range, mu, std_dev), 'LineWidth', 1.5);
    plot(x_range, betapdf(x_range, a_beta, b_beta), 'LineWidth', 1.5);
    plot(x_range, versatile_pdf_func(popt, x_range), 'r--', 'LineWidth', 2);
    
    xlabel('Probabilistic wind power output'); ylabel('PDF');
    legend('Actual (Binned)', 'Gaussian', 'Beta', 'Versatile', 'Location', 'Best');
    title(['Comparison of PDF Fitting - Bin ', bin_num]);
    grid on;
    saveas(gcf, pdf_name);
    close(gcf);

    % --- RMSE 评估 ---
    n_o = 100;
    eval_points = linspace(1/n_o, 1, n_o)';
    actual_cdf_at_eval = interp1(data_sorted, y_actual_cdf, eval_points, 'linear', 'extrap');
    
    rmse_g = sqrt(mean((actual_cdf_at_eval - normcdf(eval_points, mu, std_dev)).^2));
    rmse_b = sqrt(mean((actual_cdf_at_eval - betacdf(eval_points, a_beta, b_beta)).^2));
    rmse_v = sqrt(mean((actual_cdf_at_eval - versatile_cdf_func(popt, eval_points)).^2));

    % --- 写入 TXT ---
    fid = fopen(txt_name, 'w');
    fprintf(fid, '--- Fitting Parameters ---\n');
    fprintf(fid, 'Gaussian: mu=%.4f, std=%.4f\n', mu, std_dev);
    fprintf(fid, 'Beta: a=%.4f, b=%.4f\n', a_beta, b_beta);
    fprintf(fid, 'Versatile: a=%.4f, b=%.4f, c=%.4f\n\n', a_v, b_v, c_v);
    fprintf(fid, '--- RMSE Evaluation ---\n');
    fprintf(fid, 'Gaussian RMSE:  %.5f\n', rmse_g);
    fprintf(fid, 'Beta RMSE:      %.5f\n', rmse_b);
    fprintf(fid, 'Versatile RMSE: %.5f (Optimal)\n', rmse_v);
    fclose(fid);
    
    fprintf('Bin %s: finished fitting.\n', bin_num);
end